java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -0,0 +1,643 @@
+        return ((JavaSymbol.MethodJavaSymbol)s).completeSignature();
+      } else {
+        throw new IllegalArgumentException();
+      }
+    }
+
+    public static boolean isString(Type type) {
+      return type.is("java.lang.String");
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol) {
+      return taintSourceFor(symbol, Collections.emptyList());
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol, List<TaintSource> arguments) {
+      if (symbol.isMethodSymbol()) {
+        MethodSymbol method = (MethodSymbol)symbol;
+
+        if (arguments.stream().anyMatch(ts -> ts.canBeTainted())) {
+          methodCalls.add(new DirectTaintSource(fullyQualify(symbol), arguments));
+        }
+
+        if (isString(method.returnType().type())) {
+          return new DirectTaintSource(fullyQualify(symbol), arguments);
+        } else {
+          return new TaintFreeSource();
+        }
+      }
+
+      if (symbol.isVariableSymbol() && symbol.owner().isTypeSymbol()) {
+        // Fields are not supported
+        return new TaintFreeSource();
+      }
+
+      return taintSources.computeIfAbsent(
+        symbol,
+        s -> new DirectTaintSource(fullyQualify(s), arguments));
+    }
+
+    public void put(Symbol symbol, TaintSource ts) {
+      taintSources.put(symbol, ts);
+    }
+
+    @Override
+    public int hashCode() {
+      return src.hashCode() +
+        13 * taintSources.hashCode() +
+        19 * methodCalls.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj.getClass().equals(this.getClass()))) {
+        return false;
+      }
+
+      TaintProgramState other = (TaintProgramState) obj;
+
+      return src.equals(other.src) &&
+        taintSources.equals(other.taintSources) &&
+        methodCalls.equals(other.methodCalls);
+    }
+
+  }
+
+  public static TaintSummary computeTaintConditions(ScannedFile src, CFG cfg) {
+    Set<TaintSource> result = new HashSet<>();
+    Set<DirectTaintSource> methodCalls = new HashSet<>();
+
+    boolean returnsVoid = cfg.methodSymbol().returnType().type().isVoid();
+
+    Multimap<Block, TaintProgramState> visited = HashMultimap.create();
+    Multimap<Block, TaintProgramState> workList = HashMultimap.create();
+
+    Block entry = cfg.entry();
+    TaintProgramState entryState = new TaintProgramState(src);
+    workList.put(entry, entryState);
+
+    while (!workList.isEmpty()) {
+      Block block = workList.keys().iterator().next();
+      TaintProgramState state = workList.get(block).iterator().next();
+      visited.put(block, state);
+      if (!workList.remove(block, state)) {
+        throw new IllegalStateException();
+      }
+
+      BlockVisitor visitor = new BlockVisitor(state.dup());
+      visitor.visit(block);
+      TaintProgramState exitState = visitor.state;
+
+      methodCalls.addAll(exitState.methodCalls);
+
+      if (block.successors().contains(cfg.exitBlock())) {
+        if (!returnsVoid) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1066)